src/extract_topics.cpp

```diff
@@ -0,0 +1,124 @@
+#include <fstream>
+#include <optional>
+#include <string>
+
+#include "pisa/util/string_util.hpp"
+
+#include "CLI/CLI.hpp"
+
+namespace detail {
+
+static std::string const TOP       = "<top>";
+static std::string const TOP_END   = "</top>";
+static std::string const NUM       = "<num>";
+static std::string const NUM_ATT   = "Number:";
+static std::string const NUM_END   = "</num>";
+static std::string const TITLE     = "<title>";
+static std::string const TITLE_END = "</title>";
+static std::string const DESC      = "<desc>";
+static std::string const DESC_ATT  = "Description:";
+static std::string const DESC_END  = "</desc>";
+static std::string const NARR      = "<narr>";
+static std::string const NARR_ATT  = "Narrative:";
+static std::string const NARR_END  = "</narr>";
+
+struct Topic {
+    std::string num;
+    std::string title;
+    std::string desc;
+    std::string narr;
+};
+
+void consume(std::istream &is, std::string const &token, bool strict = true) {
+    is >> std::ws;
+    for (auto pos = token.begin(); pos != token.end(); ++pos) {
+        if (is.get() != *pos) {
+            is.unget();
+            for (auto rpos = std::reverse_iterator(pos); rpos != token.rend(); ++rpos) {
+                is.putback(*rpos);
+            }
+            if (strict) {
+                throw std::runtime_error("Could not consume tag: " + token);
+            }
+            break;
+        }
+    }
+}
+
+template <typename Pred>
+std::ostream &read_until(std::istream &is, Pred pred, std::ostream &os) {
+    is >> std::ws;
+    while (not is.eof()) {
+        if (is.peek() == std::istream::traits_type::eof() or pred(is.peek())) {
+            break;
+        }
+        os.put(is.get());
+    }
+    return os;
+}
+
+std::optional<Topic> next_topic(std::ifstream &is) {
```

Comment:
Can we extract this to a header and write unit tests?

Reply:
Yes! Will do.